### PR TITLE
[Train] Fix HF Trainer with `DatasetIterator`, handle `device_map`

### PR DESCRIPTION
**Pull request by Yard1**
## Why are these changes needed?

This PR switches to the `DatasetIterator` in Hugging Face Trainer code to avoid deprecation warnings and makes sure that the `device_map` argument in `HuggingFacePredictor` is handled properly.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/huggingface/huggingface_predictor.py [hunk 1]
```diff
@@ -74,7 +74,8 @@ def __init__(
                 "prediction will only use CPU. Please consider explicitly "
                 "setting `HuggingFacePredictor(use_gpu=True)` or "
                 "`batch_predictor.predict(ds, num_gpus_per_worker=1)` to "
-                "enable GPU prediction."
+                "enable GPU prediction. Ignore if you have set `device` or "
+                "`device_map` arguments in the `pipeline` manually."
```
Comment (bveeramani) [inline python/ray/train/huggingface/huggingface_predictor.py:78]: Would it be possible to check for this and not log a warning?
> if you have set `device` or `device_map` arguments in the `pipeline` manually

**Review by bveeramani — APPROVED** [review]
LGTM

File: python/ray/train/huggingface/huggingface_predictor.py (hunk 1 above)
Comment (Yard1) [reply python/ray/train/huggingface/huggingface_predictor.py:78]: it is possible for `device`, but `device_map` would require usage of private APIs

Comment (bveeramani) [reply python/ray/train/huggingface/huggingface_predictor.py:78]: Got it. Makes sense

File: python/ray/train/huggingface/_huggingface_utils.py [hunk 2]
```diff
@@ -84,7 +84,7 @@ def process_dataset_for_hf(dataset: Dataset) -> "IterableDataset":
     ).with_format("torch")
 
     try:
-        dataset_length = dataset.count()
+        dataset_length = dataset._base_dataset_or_pipeline.count()
```
Comment (amogkam) [inline python/ray/train/huggingface/_huggingface_utils.py:87]: With datasets streaming by default, `count()` will trigger full execution of the dataset.

Is length required here, or can we go with the pipelined case as the default?

Comment (Yard1) [reply python/ray/train/huggingface/_huggingface_utils.py:87]: Having the length would make for best UX. How about we discuss that in a followup?

Comment (amogkam) [reply python/ray/train/huggingface/_huggingface_utils.py:87]: ok can we not use private attributes from DatasetIterator?

Comment (Yard1) [reply python/ray/train/huggingface/_huggingface_utils.py:87]: How can I obtain the count in that case?

Comment (amogkam) [reply python/ray/train/huggingface/_huggingface_utils.py:87]: ok that's fine, but let's make sure to follow up

**Review by amogkam — APPROVED** [review]

File: python/ray/train/huggingface/_huggingface_utils.py (hunk 2 above)
Comment (Yard1) [reply python/ray/train/huggingface/_huggingface_utils.py:87]: I'll make an issue

**Pull request merged**